MovieLense Example for MultiHot

bschifferer opened the pull request:
Added example for multi-hot based on MovieLense and TensorFlow.
Added mini example in Categorify op documentation

ToDo: Add PyTorch

bschifferer commented:
run tests

bschifferer commented on nvtabular/framework_utils/tensorflow/layers/embedding.py:
```diff
@@ -374,3 +374,66 @@ def get_config(self):
         }
+
+
+class SparseEmbeddingLookup(tf.keras.layers.Layer):
```
Thanks for the comment. I was not aware, DenseFeature supports this. I updated the example

benfred commented on examples/movielense_multihot_example.ipynb:
```diff
@@ -0,0 +1,1309 @@
+    }
+   ],
+   "source": [
+    "ratings = cudf.read_csv('/raid/data/ml/ml-25m/ratings.csv')\n",
```
We should use the directory variable (BASE_DIR) instead of hardcoding the path:
```suggestion
    "ratings = cudf.read_csv(os.path.join(BASE_DIR, "ml-25m", ratings.csv'))\n",
```

benfred commented on examples/movielense_multihot_example.ipynb:
```diff
@@ -0,0 +1,1309 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we take a look on our data loader and how the data is represented as tensors. The NVTabular data loader are initialized as usually and we specify both single-hot and multi-hot categorical features as cat_names. The data loader will automatically recognize the single/multi-hot data loader and represent them accordingly."
```
```suggestion
    "First, we take a look on our data loader and how the data is represented as tensors. The NVTabular data loader are initialized as usually and we specify both single-hot and multi-hot categorical features as cat_names. The data loader will automatically recognize the single/multi-hot columns and represent them accordingly."
```

benfred commented on examples/movielense_multihot_example.ipynb:
```diff
@@ -0,0 +1,1309 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "NVTabular implemented a custom TensorFlow layer `layers.DenseFeatures`, which takes as an input the different `tf.Keras.Input` and pre-initialized `tf.feature_column` and automatically concatenate them into a flat tensor. In the case of multi-hot categorical features, `DenseFeatures` organizes the inputs `__values` and `__nnzs` to define a `RaggedTensor` and combine them. `DenseFeatures` can ...
```
```suggestion
    "NVTabular implemented a custom TensorFlow layer `layers.DenseFeatures`, which takes as an input the different `tf.Keras.Input` and pre-initialized `tf.feature_column` and automatically concatenate them into a flat tensor. In the case of multi-hot categorical features, `DenseFeatures` organizes the inputs `__values` and `__nnzs` to define a `RaggedTensor` and combine them. `DenseFeatures` can handle numeric inputs, as well, but MovieLens does not provide numerical input features."
```

benfred commented on examples/movielense_multihot_example.ipynb:
```diff
@@ -0,0 +1,1309 @@
+   "source": [
+    "<img src=\"http://developer.download.nvidia.com/compute/machine-learning/frameworks/nvidia_logo.png\" style=\"width: 90px; float: right;\">\n",
+    "\n",
+    "# MovieLense: NVTabular for MultiHot Categories on MovieLense25M\n",
```
```suggestion
    "# MovieLens: NVTabular for MultiHot Categories on MovieLens25M\n",
```

benfred reviewed (changes requested):
Thanks for updating to use DenseFeatures!

Can you rename the file from movielense_multihot_example.ipynb  to movielens_multihot_example.ipynb ? (there is no final 'E' in movielens)

bschifferer commented:
> Thanks for updating to use DenseFeatures!
> 
> Can you rename the file from movielense_multihot_example.ipynb to movielens_multihot_example.ipynb ? (there is no final 'E' in movielens)

thanks - I renamed the notebook and updated all suggestions

benfred commented:
rerun tests

benfred commented:
rerun tests
